### PR TITLE
Fix cross-loop blocking with imperfect nests

### PR DESCRIPTION
**Pull request by FabioLuporini**

### CODE REVIEW

File: devito/ir/clusters/analysis.py [hunk 1]
```diff
@@ -75,6 +75,10 @@ class Parallelism(Detector):
     """
 
     def _callback(self, clusters, d, prefix):
+        # Rule out if non-unitary increment Dimension (e.g., `t0=(time+1)%2`)
+        if any(c.sub_iterators.get(d) for c in clusters):
+            return SEQUENTIAL
+
         # All Dimensions up to and including `i-1`
```
Comment (georgebisbas) [inline devito/ir/clusters/analysis.py:82]: better docstring>?

File: devito/passes/clusters/blocking.py [hunk 2]
```diff
@@ -116,24 +116,28 @@ def decompose(ispace, d, block_dims):
...
+    # Add more relations
+    for n, i in enumerate(ispace.intervals):
+        if i.dim is d:
             continue
-        for bd in block_dims:
-            if bd._defines & i.dim._defines:
-                break
-            if len(i.dim._defines) > len(bd._defines):
+        elif i.dim.is_Incr:
+            # Make sure IncrDimensions on the same level stick next to each other.
```
Comment (georgebisbas) [inline devito/passes/clusters/blocking.py:132]: Maybe improve docstring?

File: tests/test_dle.py [hunk 3]
```diff
@@ -313,6 +313,51 @@ def test_cache_blocking_imperfect_nest(blockinner):
     assert np.all(v.data == v1.data)
 
 
+@pytest.mark.parametrize("blockinner", [False, True])
+def test_cache_blocking_imperfect_nest_v2(blockinner):
+    """
+    Test that a non-perfect Iteration nest is blocked correctly. This
```
Comment (georgebisbas) [inline tests/test_dle.py:319]: Maybe use either imperfect or on-perfect for consistency? I guess imperfect is the better?

File: devito/ir/clusters/analysis.py (hunk 1 above)
Comment (FabioLuporini) [reply devito/ir/clusters/analysis.py:82]: not part of this PR, but you have a point

File: devito/passes/clusters/blocking.py (hunk 2 above)
Comment (FabioLuporini) [reply devito/passes/clusters/blocking.py:132]: I think it's fairly clear but you need to know what this function does and how the blocking implementation works. At least a little

File: tests/test_dle.py (hunk 3 above)
Comment (FabioLuporini) [reply tests/test_dle.py:319]: it's a test docstring, so... if I'll have to touch again this branch, I will

File: devito/passes/clusters/blocking.py (hunk 2 above)
Comment (georgebisbas) [reply devito/passes/clusters/blocking.py:132]: I mean mostly the part 'stick next to each other' ..... (?)

**Comment by FabioLuporini** [comment]
merged

**Pull request merged**